extmod/uasyncio/core.py

```diff
@@ -0,0 +1,500 @@
+    try:
+        s.connect(ai[-1])
+    except OSError as er:
+        if er.args[0] != 115: # EINPROGRESS
+            raise er
+    yield _io_queue.queue_write(s)
+    return ss, ss
+
+# Class representing a TCP stream server, can be closed and used in "async with"
+class Server:
+    async def __aenter__(self):
+        return self
+    async def __aexit__(self, exc_type, exc, tb):
+        self.close()
+        await self.wait_closed()
+    def close(self):
+        self.task.cancel()
+    async def wait_closed(self):
+        await self.task
+    async def _serve(self, cb, host, port, backlog):
+        try:
+            import usocket as socket
+        except ImportError:
+            import socket
+        ai = socket.getaddrinfo(host, port)[0] # TODO this is blocking!
+        s = socket.socket()
+        s.setblocking(False)
+        s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        s.bind(ai[-1])
+        s.listen(backlog)
+        self.task = cur_task
+        # Accept incoming connections
+        while True:
+            try:
+                yield _io_queue.queue_read(s)
+            except CancelledError:
+                # Shutdown server
+                s.close()
+                return
+            s2, addr = s.accept()
+            s2.setblocking(False)
+            s2s = Stream(s2, {'peername': addr})
+            create_task(cb(s2s, s2s))
+
+# Helper function to start a TCP stream server, running as a new task
+# TODO could use an accept-callback on socket read activity instead of creating a task
+async def start_server(cb, host, port, backlog=5):
+    s = Server()
+    create_task(s._serve(cb, host, port, backlog))
+    return s
+
+################################################################################
+# Main run loop
+
+# Queue of Task instances
+_queue = Queue()
+
+# Task queue and poller for stream IO
+_io_queue = IOQueue()
+
+# Keep scheduling tasks until there are none left to schedule
+def run_until_complete(main_task=None):
+    global cur_task
+    excs_all = (CancelledError, Exception) # To prevent heap allocation in loop
+    excs_stop = (CancelledError, StopIteration) # To prevent heap allocation in loop
+    while True:
+        # Wait until the head of _queue is ready to run
+        dt = 1
+        while dt > 0:
+            dt = -1
+            t = _queue.peek()
+            if t:
+                # A task waiting on _queue; "ph_key" is time to schedule task at
+                dt = max(0, ticks_diff(t.ph_key, ticks()))
+            elif not _io_queue.map:
+                # No tasks can be woken so finished running
+                return
+            #print('(poll {})'.format(dt), len(_io_queue.map))
+            _io_queue.wait_io_event(dt)
+
+        # Get next task to run and continue it
+        t = _queue.pop_head()
+        cur_task = t
+        try:
+            # Continue running the coroutine, it's responsible for rescheduling itself
+            if not t.data:
```

Comment:
Not sure if this is the right fix.  It will impact performance for a very uncommon case.  I'd really like to get a test case that triggers this problem, then see how to fix it.